code/modules/food_and_drinks/recipes/soup_mixtures.dm

```diff
@@ -62,19 +62,19 @@
 	if(!length(required_ingredients))
 		return TRUE
 
-	// This is very unoptimized for something ran every handle-reaction for every soup recipe.
-	// Look into ways for improving this, cause bleh
+	// See if we fulfill all reqs
 	var/list/reqs_copy = required_ingredients.Copy()
 	for(var/obj/item/ingredient as anything in pot.added_ingredients)
-		// See if we fulfill all reqs
-		for(var/ingredient_type in required_ingredients)
-			if(!istype(ingredient, ingredient_type))
-				continue
+		var/ingredient_type = ingredient.type
+
+		var/ingredient_count = reqs_copy[ingredient_type]
+		if(ingredient_count)
```

Comment:
These are not equivalent; the check needs to allow for subtypes when checking ingredients. That's why it is currently using `istype`.

Reply:
Any other way to optimize this then? I'm out of ideas

Reply:
I think I might have a idea, will have to test tomorrow

Reply:
I had an idea which involved building a list of possible reactions as ingredients are added but it'd involve a lot of code rewrite.

Reply:
That is a lot of work true. A type cache would also be too large and a waste since the reaction lasts only a short time and then gets deleted anyway.

I'm thinking of mimicking the behavior of a typecache in a scaled down way . We'll see tom.